pkg/runtime/wfengine/wfengine.go

```diff
@@ -145,41 +148,40 @@ func (wfe *WorkflowEngine) SetActorReminderInterval(interval time.Duration) {
 }
 
 func (wfe *WorkflowEngine) Start(ctx context.Context) error {
-	var err error
-	wfe.startedOnce.Do(func() {
-		if wfe.actorRuntime == nil {
-			err = errors.New("actor runtime is not configured")
-			return
-		} else if wfe.executor == nil {
-			err = errors.New("grpc executor is not yet configured")
-			return
-		}
+	// Start could theoretically get called by multiple goroutines concurrently
+	wfe.startMutex.Lock()
```

Comment:
If the problem you're trying to solve is make sure this isn't started more than once, I'd use a simpler semaphore:

```go
// property m.isStarted is sync.Bool{}

if !m.isStarted.CompareAndSwap(false, true) {
    // If we're here, it's already running
    return nil
}
```

Then to stop this:

```go
m.isStarted.Store(false)
```

This doesn't require a lock, which means that concurrent callers will get a result (nil) right away rather than having to wait for the first caller's invocation of `Start()` to complete.

Reply:
Thanks for sharing this valuable tip! I was not aware of `sync.Bool`. However, I think I would actually prefer to keep the mutex in this case. That way I can ensure that the `Start()` method always returns only after the engine has actually started. This code path will be infrequently hit, so I'm not worried about the perf impact. :)

Reply:
this is a perfect fit for [singleflights](https://pkg.go.dev/golang.org/x/sync/singleflight), but this is a nit